### PR TITLE
Generate OpenAPI title

### PR DESCRIPTION
**Pull request by vmunishwar**
Generates an Openapi title key from schema/title annotations
so that Configuration Consumers see a short explanation of the key
Openapi title key will appear as a first key except for null_key which will be taken care in [#551 ](https://github.com/vmware-tanzu/carvel-ytt/issues/551)

### CODE REVIEW

File: pkg/schema/schema.go [hunk 1]
```diff
@@ -118,6 +118,11 @@ func getType(node yamlmeta.Node) (Type, error) {
 			typeOfValue.SetDescription(desc.description)
 		}
 	}
+	for _, ann := range docAnns {
+		if titleAnn, ok := ann.(*TitleAnnotation); ok {
+			typeOfValue.SetTitle(titleAnn.title)
+		}
+	}
```
Comment (pivotaljohn) [inline pkg/schema/schema.go]: We now have two cases within the same loop.

How about a single `switch` within the `for` with a `case` for each documentation annotation?

File: pkg/schema/openapi.go [hunk 2]
```diff
@@ -57,11 +58,15 @@ func (o *OpenAPIDocument) calculateProperties(schemaVal interface{}) *yamlmeta.M
...
-			{Key: typeProp, Value: "object"},
-			{Key: "additionalProperties", Value: false},
-		}}
-
+		var property yamlmeta.Map
+		if typedValue.GetTitle() != "" {
+			item := []*yamlmeta.MapItem{
+				{Key: titleProp, Value: typedValue.GetTitle()},
+			}
+			property.Items = append(property.Items, item...)
```
Comment (pivotaljohn) [inline pkg/schema/openapi.go]: `item` need not be a slice of `yamlmeta.MapItem`, we know we're adding exactly one. That makes declaring an array/slice "accidental complexity."

Let's make it a simple pointer to a map item, instead.

File: pkg/schema/annotations.go [hunk 3]
```diff
@@ -297,6 +358,15 @@ func collectDocumentationAnnotations(node yamlmeta.Node) ([]Annotation, error) {
...
 	}
+	for _, annotation := range []template.AnnotationName{AnnotationTitle} {
+		ann, err := processOptionalAnnotation(node, annotation, nil)
+		if err != nil {
+			return nil, err
+		}
+		if ann != nil {
+			anns = append(anns, ann)
+		}
+	}
```
Comment (pivotaljohn) [inline pkg/schema/annotations.go]: We've got duplicate code, here. The `for` loop is iterating over the annotation names.

I think you meant to add `AnnotationTitle` to the list of `AnnotationName`s.

**Review by pivotaljohn — CHANGES_REQUESTED** [review]
Overall, looking good. 👍🏻 

A few tune-ups required.

**Comment by gcheadle-vmware** [comment]
From my perspective, these changes look really good! You extend the existing patterns that are being used to add descriptions, and applied them to adding a title. Your error cases read concisely and are actionable. There were only the few things that John pointed out, that I agree would improve readability and simplicity. Great work!

File: pkg/schema/schema.go (hunk 1 above)
Comment (vmunishwar) [reply pkg/schema/schema.go]: Thanks for the review. I have made the changes to the code.

File: pkg/schema/openapi.go (hunk 2 above)
Comment (vmunishwar) [reply pkg/schema/openapi.go]: Made the changes. The code looks simpler and consistent now.

File: pkg/schema/annotations.go (hunk 3 above)
Comment (vmunishwar) [reply pkg/schema/annotations.go]: Good catch. I have updated the code.

File: pkg/cmd/template/schema_inspect_test.go [hunk 4]
```diff
@@ -624,6 +623,84 @@ components:
 	})
 }
 
+func TestSchemaInspect_annotation_adds_title_key(t *testing.T) {
```
Comment (cari-lynn) [inline pkg/cmd/template/schema_inspect_test.go]: Nice job on these tests. For organization, let's move the subtests from ` TestSchemaInspect_annotation_adds_title_key` to the test above: `TestSchemaInspect_annotation_adds_key`. That test tests adding 'documenting' keys when an annotation that provides documentation information is present. As opposed to the other schema annotations we have (`@schema/default`, `@schema/type`, `@schema/nullable`) add constraints to the schema.

**Review by cari-lynn — CHANGES_REQUESTED** [review]
Great changes. I did some manual testing as well using multiple schema annotations per node, and tuples as arguments, and all worked great.

**Review by pivotaljohn — APPROVED** [review]
After addressing the item that @cari-lynn brought up (so much love on this PR!), this all LGTM, @vmunishwar.

After you address her feedback and check back in with her, I'd say feel free to merge. 👍🏻

File: pkg/cmd/template/schema_inspect_test.go (hunk 4 above)
Comment (vmunishwar) [reply pkg/cmd/template/schema_inspect_test.go]: Thanks for the review. I have moved the subtest as suggested and committed the code.

**Review by cari-lynn — APPROVED** [review]

**Pull request merged**